# Generated by Django 2.2.26 on 2022-09-09 18:30

from django.conf import settings
from django.db import migrations, models
import django.db.models.deletion
import django.db.models.manager
import mibios.models
import mibios_glamr.fields


class Migration(migrations.Migration):

    initial = True

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='Dataset',
            fields=[
                ('id', mibios.models.AutoField(primary_key=True, serialize=False)),
                ('short_name', models.CharField(blank=True, default=None, help_text='a short name or description, for internal use, not (necessarily) for public display', max_length=64, null=True, unique=True)),
                ('dataset_id', models.PositiveIntegerField(help_text='GLAMR accession to data set/study/project', unique=True, verbose_name='Study ID')),
                ('bioproject', models.CharField(blank=True, default='', max_length=100)),
                ('jgi_project', models.CharField(blank=True, default='', max_length=100)),
                ('gold_id', models.CharField(blank=True, default='', max_length=100)),
                ('scheme', models.CharField(blank=True, default='', max_length=512, verbose_name='location and sampling scheme')),
                ('material_type', models.CharField(blank=True, default='', max_length=128)),
                ('water_bodies', models.CharField(blank=True, default='', help_text='list or description of sampled bodies of water', max_length=256)),
                ('primers', models.CharField(blank=True, default='', max_length=64)),
                ('sequencing_target', models.CharField(blank=True, default='', max_length=64)),
                ('sequencing_platform', models.CharField(blank=True, default='', max_length=64)),
                ('size_fraction', models.CharField(blank=True, default='', help_text='e.g.: >0.22µm or 0.22-1.6µm', max_length=32)),
                ('note', models.TextField(blank=True, default='')),
            ],
            options={
                'default_manager_name': 'objects',
            },
        ),
        migrations.CreateModel(
            name='Sample',
            fields=[
                ('id', mibios.models.AutoField(primary_key=True, serialize=False)),
                ('tracking_id', models.CharField(blank=True, default=None, help_text='internal uniform hex id', max_length=32, null=True, unique=True)),
                ('sample_id', models.CharField(blank=True, default=None, help_text='internal sample accession', max_length=256, null=True, unique=True)),
                ('sample_name', models.CharField(help_text='sample ID or name as given by study', max_length=64)),
                ('sample_type', models.CharField(blank=True, choices=[('amplicon', 'amplicon'), ('metagenome', 'metagenome'), ('metatranscriptome', 'metatranscriptome')], default=None, max_length=32, null=True)),
                ('has_paired_data', models.BooleanField(blank=True, default=None, null=True)),
                ('sra_accession', models.CharField(blank=True, default='', help_text='SRA accession', max_length=16)),
                ('amplicon_target', models.CharField(blank=True, default='', max_length=16)),
                ('fwd_primer', models.CharField(blank=True, default='', max_length=32)),
                ('rev_primer', models.CharField(blank=True, default='', max_length=32)),
                ('meta_data_loaded', models.BooleanField(default=False, help_text='meta data successfully loaded')),
                ('metag_pipeline_reg', models.BooleanField(default=False, help_text='is registered in metagenomic pipeline, has tracking ID')),
                ('contig_fasta_loaded', models.BooleanField(default=False, help_text='contig fasta data loaded')),
                ('gene_fasta_loaded', models.BooleanField(default=False, help_text='gene fasta data loaded')),
                ('contig_abundance_loaded', models.BooleanField(default=False, help_text='contig abundance/rpkm data data loaded')),
                ('gene_abundance_loaded', models.BooleanField(default=False, help_text='gene abundance/rpkm data loaded')),
                ('gene_alignment_hits_loaded', models.BooleanField(default=False, help_text='gene alignment hits to UniRef100 loaded')),
                ('binning_ok', models.BooleanField(default=False, help_text='Binning data loaded')),
                ('checkm_ok', models.BooleanField(default=False, help_text='Binning stats loaded')),
                ('genes_ok', models.BooleanField(default=False, help_text='Gene data and coverage loaded')),
                ('proteins_ok', models.BooleanField(default=False, help_text='Protein data loaded')),
                ('tax_abund_ok', models.BooleanField(default=False, help_text='Taxon abundance data loaded')),
                ('func_abund_ok', models.BooleanField(default=False, help_text='Function abundance data loaded')),
                ('comp_abund_ok', models.BooleanField(default=False, help_text='Compound abundance data loaded')),
                ('analysis_dir', models.CharField(blank=True, default=None, help_text='path to results of analysis, relative to OMICS_DATA_ROOT', max_length=256, null=True)),
                ('read_count', models.PositiveIntegerField(blank=True, default=None, help_text='number of reads (post-QC, fwd+rev) used for assembly mapping', null=True)),
                ('reads_mapped_contigs', models.PositiveIntegerField(blank=True, default=None, help_text='number of reads mapped to contigs', null=True)),
                ('reads_mapped_genes', models.PositiveIntegerField(blank=True, default=None, help_text='number of reads mapped to genes', null=True)),
                ('project_id', models.CharField(blank=True, default='', help_text='Project accession, e.g. NCBI bioproject', max_length=16)),
                ('biosample', models.CharField(blank=True, default='', max_length=16)),
                ('geo_loc_name', models.CharField(blank=True, default='', max_length=256)),
                ('gaz_id', models.CharField(blank=True, default='', max_length=16, verbose_name='GAZ id')),
                ('latitude', models.CharField(blank=True, default='', max_length=16)),
                ('longitude', models.CharField(blank=True, default='', max_length=16)),
                ('collection_timestamp', models.DateTimeField(blank=True, default=None, null=True)),
                ('noaa_site', models.CharField(blank=True, default='', max_length=16, verbose_name='NOAA Site')),
                ('env_broad_scale', models.CharField(blank=True, default='', max_length=64)),
                ('env_local_scale', models.CharField(blank=True, default='', max_length=64)),
                ('env_medium', models.CharField(blank=True, default='', max_length=64)),
                ('modified_or_experimental', models.BooleanField(default=False)),
                ('depth', models.CharField(blank=True, default='', max_length=16)),
                ('depth_sediment', models.CharField(blank=True, default='', max_length=16)),
                ('size_frac_up', models.CharField(blank=True, default='', max_length=16)),
                ('size_frac_low', models.CharField(blank=True, default='', max_length=16)),
                ('ph', models.CharField(blank=True, default='', max_length=8, verbose_name='pH')),
                ('temp', models.CharField(blank=True, default='', max_length=8)),
                ('calcium', models.CharField(blank=True, default='', max_length=8)),
                ('potassium', models.CharField(blank=True, default='', max_length=8)),
                ('magnesium', models.CharField(blank=True, default='', max_length=8)),
                ('ammonium', models.CharField(blank=True, default='', max_length=8)),
                ('nitrate', models.CharField(blank=True, default='', max_length=8)),
                ('phosphorus', models.CharField(blank=True, default='', max_length=8)),
                ('diss_oxygen', models.CharField(blank=True, default='', max_length=8)),
                ('conduc', models.CharField(blank=True, default='', max_length=16)),
                ('secci', models.CharField(blank=True, default='', max_length=8)),
                ('turbidity', models.CharField(blank=True, default='', max_length=8)),
                ('part_microcyst', models.CharField(blank=True, default='', max_length=8)),
                ('diss_microcyst', models.CharField(blank=True, default='', max_length=8)),
                ('ext_phyco', models.CharField(blank=True, default='', max_length=8)),
                ('ext_microcyst', models.CharField(blank=True, default='', max_length=8)),
                ('ext_anatox', models.CharField(blank=True, default='', max_length=8)),
                ('chlorophyl', models.CharField(blank=True, default='', max_length=8)),
                ('diss_phosp', models.CharField(blank=True, default='', max_length=8)),
                ('soluble_react_phosp', models.CharField(blank=True, default='', max_length=8)),
                ('ammonia', models.CharField(blank=True, default='', max_length=8)),
                ('nitrate_nitrite', models.CharField(blank=True, default='', max_length=8)),
                ('urea', models.CharField(blank=True, default='', max_length=8)),
                ('part_org_carb', models.CharField(blank=True, default='', max_length=8)),
                ('part_org_nitro', models.CharField(blank=True, default='', max_length=8)),
                ('diss_org_carb', models.CharField(blank=True, default='', max_length=8)),
                ('col_dom', models.CharField(blank=True, default='', max_length=8)),
                ('suspend_part_matter', models.CharField(blank=True, default='', max_length=8)),
                ('suspend_vol_solid', models.CharField(blank=True, default='', max_length=8)),
                ('microcystis_count', models.PositiveIntegerField(blank=True, default=None, null=True)),
                ('planktothris_count', models.PositiveIntegerField(blank=True, default=None, null=True)),
                ('anabaena_d_count', models.PositiveIntegerField(blank=True, default=None, null=True)),
                ('cylindrospermopsis_count', models.PositiveIntegerField(blank=True, default=None, null=True)),
                ('notes', models.CharField(blank=True, default='', max_length=512)),
                ('dataset', models.ForeignKey(blank=True, default=None, null=True, on_delete=django.db.models.deletion.SET_NULL, to=settings.OMICS_DATASET_MODEL)),
            ],
            options={
                'default_manager_name': 'objects',
            },
        ),
        migrations.CreateModel(
            name='Reference',
            fields=[
                ('id', mibios.models.AutoField(primary_key=True, serialize=False)),
                ('short_reference', models.CharField(help_text='short reference', max_length=128)),
                ('authors', models.CharField(blank=True, default='', help_text='author listing', max_length=2048)),
                ('title', models.CharField(blank=True, default='', max_length=512)),
                ('abstract', models.TextField(blank=True, default='')),
                ('key_words', models.CharField(blank=True, default='', max_length=128)),
                ('publication', models.CharField(blank=True, default='', max_length=128)),
                ('doi', mibios_glamr.fields.OptionalURLField(blank=True, default=None, null=True, unique=True)),
            ],
            options={
                'unique_together': {('short_reference', 'doi')},
            },
            managers=[
                ('loader', django.db.models.manager.Manager()),
            ],
        ),
        migrations.AddField(
            model_name='dataset',
            name='reference',
            field=models.ForeignKey(blank=True, default=None, null=True, on_delete=django.db.models.deletion.SET_NULL, to='mibios_glamr.Reference'),
        ),
    ]
