# Generated by Django 3.2.16 on 2022-12-20 20:44

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('omics', '0001_initial'),
    ]

    operations = [
        migrations.RemoveField(
            model_name='taxonabundance',
            name='sum_gene_rpkm',
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='count_contig',
            field=models.PositiveIntegerField(default=None, verbose_name='number of contigs'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='count_gene',
            field=models.PositiveIntegerField(default=None, verbose_name='number of genes'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='len_contig',
            field=models.PositiveIntegerField(default=None, verbose_name='total length of all contigs'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='len_gene',
            field=models.PositiveIntegerField(default=None, verbose_name='total length of all genes'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='mean_fpkm_contig',
            field=models.FloatField(default=None, verbose_name='aggregated fpkm across all contigs'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='mean_fpkm_gene',
            field=models.FloatField(default=None, verbose_name='aggregated fpkm across all genes'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='norm_frags_contig',
            field=models.FloatField(default=None, verbose_name='sum of fragments mapping to contigs, normalized'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='norm_frags_gene',
            field=models.FloatField(default=None, verbose_name='sum of fragments mapping to genes, normalized'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='norm_reads_contig',
            field=models.FloatField(default=None, verbose_name='sum of reads mapping to contigs, normalized'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='norm_reads_gene',
            field=models.FloatField(default=None, verbose_name='sum of reads mapping to genes, normalized'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='wmedian_fpkm_contig',
            field=models.FloatField(default=None, verbose_name='weighted median fpkm across all contigs'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='taxonabundance',
            name='wmedian_fpkm_gene',
            field=models.FloatField(default=None, verbose_name='weighted median fpkm across all genes'),
            preserve_default=False,
        ),
    ]
