# Generated by Django 4.2.19 on 2025-03-05 16:10

# from django.conf import settings
from django.db import migrations, models

# from mibios.omics.tracking import registry as job_registry


def migrate_data_fwd_1(apps, schema_editor):
    """
    Add tracking for samples with contig_abundance_loaded

    Tracking timestamps will reflect the date/time of migration.
    """
    # The commented-out code below would not run inside a migration for two
    # reasons.  First, Sample's status as swappable model somehow result in
    # Sample not having the flag fields anymore after the corresponding glamr
    # app's migrations (should be 0028) runs.  Second, job.fake_run() depends
    # on an instance method of the Sample model which is not available in the
    # migration.  The code below or something similar should be run manually.

    # sample_app, _, sample_model = settings.OMICS_SAMPLE_MODEL.rpartition('.')
    # Sample = apps.get_model(sample_app, sample_model)
    # # get the job class
    # LoadContigAbund = job_registry.jobs['LoadContigAbund']
    # for i in Sample.objects.filter(contig_abundance_loaded=True):
    #     job = LoadContigAbund.for_sample(i)
    #     job.fake_run()
    pass


def migrate_data_fwd_2(apps, schema_editor):
    """
    Set all contig LCAs to None

    Current LCA values are obsolete.
    """
    Contig = apps.get_model('omics', 'Contig')
    num_changed = Contig.objects.exclude(lca=None).update(lca=None)
    print(f'\n    Reset lca to None on {num_changed} contigs. ')


class Migration(migrations.Migration):

    dependencies = [
        ('omics', '0023_file_pipeline_readonly'),
    ]

    operations = [
        migrations.RunPython(migrate_data_fwd_1, migrations.RunPython.noop),
        migrations.RunPython(migrate_data_fwd_2, migrations.RunPython.noop),
        migrations.RemoveField(
            model_name='sample',
            name='contig_abundance_loaded',
        ),
        migrations.RemoveField(
            model_name='sample',
            name='contig_lca_loaded',
        ),
        migrations.RemoveField(
            model_name='sample',
            name='gene_alignments_loaded',
        ),
        migrations.AlterField(
            model_name='file',
            name='filetype',
            field=models.PositiveSmallIntegerField(choices=[(1, 'metagenomic assembly, fasta format'), (2, 'metatranscriptome assembly, fasta format'), (3, 'functional abundance, csv format'), (4, 'taxonomic abundance, csv format'), (5, 'functional abundance (TPM) [csv]'), (6, 'contig abundance [csv]'), (7, 'contig taxonomy [csv]')], verbose_name='file type'),
        ),
        migrations.AlterField(
            model_name='sampletracking',
            name='flag',
            field=models.CharField(choices=[('MD', 'meta data loaded'), ('PL', 'omics pipeline registered'), ('ASM', 'assembly loaded'), ('UAB', 'reads/UR100 abundance loaded'), ('TPM', 'reads/UR100/TPM loaded'), ('TAB', 'taxa abundance loaded'), ('CAB', 'contig abundance loaded')], max_length=3),
        ),
    ]
